### ISSUE TITLE
Make PipeCD control plane support plugin-arched piped

### ISSUE THREAD

**Issue by ffjlabo** [pipe-cd/pipecd#5252]
**What would you like to be added**:

This issue tracks the progress for deprecating Platform Provider and Kind toward piped plugin architecture in pipedv1. 

The purpose is to manage application kind and platform providers on the PipeCD plugin side in the future.

Considering the impact on users, pay attention to the following points:
- In Pipedv1, platform providers, kind remain on the data and do not create logic with them. Use the value you define instead.
- minify the migration for users.

**current update process**

1. Update data for the pipedv1
  -  Add the label `kind` based on the current Application Kind to the Application.
  - Add `deployTarget`  based on the current Application Platform Provider to the Application.
2. update to pipedv1 (users can use existing application at the time)

### For Platform Provider

Instead of Platform Provider, we plan to introduce the config for the plugin and define deployTargets.

**piped config**
```yaml
apiVersion: pipecd.dev/v1beta1
kind: Piped
spec:
...
  plugin:
    - name: k8s_plugin
      port: 8081
      deployTargets:
        - name: dev
          labels: 
            env: dev
          config: # depends on plugins
            masterURL: http://cluster-dev
            kubeConfigPath: ./kubeconfig-dev
```

```golang
type PipedDeployTarget struct {
	Name   string                     `json:"name"`
	Labels map[string]string          `json:"labels,omitempty"`
	Config json.RawMessage            `json:"config"`
}
```

We also plan to deploy the app to multiple targets at once in a multicluster feature for k8s.
So, we define `DeployTargets` as an array in Application and Deployment.

**Application**

```proto
message Application {
    reserved 3;
    ...
    // TODO: Add validation for this field.
    string platform_provider = 15;
    // 
    repeated string deploy_targets = 16;
    ...
}
```

**Deployment**

```proto
message Deployment {
    reserved 4;
    ...
    // The name of platform provider where to deploy this application.
    // This must be one of the provider names registered in the piped.
    string platform_provider = 11;
    
    repeated string deploy_targets = 12;
}
```

#### For the backward compatibility

Before updating the piped, users should migrate the current data on DB.

<details>

<summary>Former idea</summary>

During the migration, there might be both platform providers and deploy targets in the piped config.
So we need to convert the platform providers to deploy targets internally.

**Refer the Platform Provider or Deploy Target**

- If the ApplicationKind is `Application`, just use `DeployTarget`
- If the ApplicationKind is old one, convert `PlatformProvider` to `DeployTarget`

```golang
func (s *PipedSpec) FindDeployTarget(name string, t model.ApplicationKind) (*PipedDeployTarget, bool) {
	// First, check the application is supported by the plugin architecture. It means that the kind is set to "Application".
	// If not, the deploy target is the platform provider.
	// For backward compatibility, the deploy target is the platform provider.
	if t != model.ApplicationKind_APPLICATION {
		p, found := s.FindPlatformProvider(name, t)
		if !found {
			return &PipedDeployTarget{}, false
		}
		return &PipedDeployTarget{
				Name:   p.Name,
				Labels: p.Labels,
				Config: p.Config,
			}, true
	}

	// If the application is supported by the plugin architecture, the deploy target is the deploy target.
	for _, dt := range s.DeployTargets {
		if dt.Name == name {
			return dt, true
		}
	}

	return &PipedDeployTarget{}, false
}
```

</details>

### For Kind

Instead of Kind, we plan to introduce the label to represent the application kind.

```yaml
apiVersion: pipecd.dev/v1beta1
kind: Application
metadata:
  labels:
    kind: KUBERNETES # <- like this
spec:
  name: myApp
```

For the builtin plugins, we define 5 labels as string.
- KUBERNETES
- ECS
- LAMBDA
- CLOUDRUN
- TERRAFORM

#### For the backward compatibility

Before updating the piped, users should migrate the current data on DB.

<details>

<summary>Former idea</summary>

We need to support both before and after creating plugin architecture for now.
So, I propose the way to decide the application kind like this.

- Define `APPLICATION` as `ApplicationKind`
- Add new method to decide the actual kind for Application and Deployment.

```proto
enum ApplicationKind {
    KUBERNETES = 0;
    TERRAFORM = 1;
    LAMBDA = 3;
    CLOUDRUN = 4;
    ECS = 5;
    APPLICATION = 6; <- new! 
}
```

```golang
func (a * Application) GetKind() string {
	// First, check the application is supported by the plugin architecture. It means that the kind is set to "Application".
	// If so, return the kind from the labels.
	if a.Kind == ApplicationKind_Application {
		return a.Labels["kind"]
	}

	// For backward compatibility, return the kind as string
	return a.Kind.String()
}
```

```golang
func (d *Deployment) GetKind() string {
	// First, check the application is supported by the plugin architecture. It means that the kind is set to "Application".
	// If so, return the kind from the labels.
	if d.Kind == ApplicationKind_Application {
		return d.Labels["kind"]
	}

	// For backward compatibility, return the kind as string
	return d.Kind.String()
}
```

</details>

## TODO

Firstly, I listed the target features to require some modifications.
We need to further investigation to decide the way to fix. 

**investigation**
- [ ] Platform Provider 
  - [x] livestatestore https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2428411125 <- @ffjlabo 
  - [x] livestatereporter https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2430816970 <- @ffjlabo 
  - [ ] executor <- maybe this is in another issue
  - [x] detector https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2430840642 <- @ffjlabo
  - [x] plan preview https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2443486833 <- @ffjlabo 
  - [x] metrics https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2449246184 <- @ffjlabo 
  - [x] UI https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2490107783 <- @khanhtc1202 
- [ ] Kind
  - [ ] executor <- maybe this is in another issue
  - [x] appconfigreporter https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2449304856 <- @ffjlabo 
  - [x] event watcher https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2449393524 <- @ffjlabo 
  - [x] livestatereporter https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2456527867 <- @ffjlabo 
  - [x] detector https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2456640194  <- @ffjlabo 
  - [x] notifier https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2456659711 <- @ffjlabo 
  - [x] plan preview https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2457063788 <- @ffjlabo 
  - [x] trigger https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2478088798 <- @ffjlabo 
  - [x] server https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2485225162 <- @ffjlabo 
  - [x] insight <- @khanhtc1202 
  - [x] metrics https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2487191963 <- @ffjlabo 
  - [x] planner https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2487239454 <- @ffjlabo 
  - [x] scheduler https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2487256861 <- @ffjlabo 
  - [x] UI <- @khanhtc1202 

**implementation**

- [x] Support Deploy Target 
  - [x] define Deploy Target
  - [ ] define method to convert to Deploy Target
- [ ] Support labels["kind"]
  - [ ] define method to convert to labels["kind"]
- [ ] Use DeployTarget instead of Platform Provider.
  - [x] livestatestore https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2428411125 
  - [x] livestatereporter https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2430816970 
  - [ ] executor <- maybe this is in another issue
  - [x] detector https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2430840642 
  - [ ] plan preview https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2443486833 
  - [ ] metrics https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2449246184 
  - [ ] UI https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2490107783 
- [ ] Use labels.["kind"] instead of Kind
  - [ ] executor <- maybe this is in another issue
  - [ ] appconfigreporter https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2449304856 
  - [ ] event watcher https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2449393524 
  - [ ] livestatereporter https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2456527867 
  - [ ] detector https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2456640194  
  - [ ] notifier https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2456659711 
  - [ ] plan preview https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2457063788 
  - [ ] trigger https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2478088798 
  - [ ] server https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2485225162 
  - [x] insight 
  - [ ] metrics https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2487191963 
  - [ ] planner https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2487239454 
  - [ ] scheduler https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2487256861 
  - [ ] UI 

**Why is this needed**:

**Comment by ffjlabo** [comment]
### Platform Provider; Livestatereporter

**The purpose of the component**

- Livestatereporter sends the application state to the Control Plane
- Currently, There are livestatereporters for each platform provider

**Modification idea**
- piped side
  - Every minute, piped does below
    - List apps by plugin name
    - Get the app livestate from the plugin via gRPC call e.g. `GetLivestate(app)`
    - Send them to the Control Plane
- plugin side
  - Get the livestates grouped by deployTarget from livestatestore and return to the piped.
- Control Plane side
  - Receive and store the livestates 

**Concern**
- There is a different way to update livestate on Control Plane between k8s and other platforms.
  - other platform (ecs, cloud run, lambda) ref for cloud run: https://github.com/pipe-cd/pipecd/blob/2f86f7dca297f7f2d402c2040a4a4e14837587ab/pkg/app/piped/livestatereporter/cloudrun/report.go#L85-L86
    - Just flush livestates to the Control Plane every minute.
  - k8s: https://github.com/pipe-cd/pipecd/blob/2f86f7dca297f7f2d402c2040a4a4e14837587ab/pkg/app/piped/livestatereporter/kubernetes/reporter.go#L99-L103
    - Flush livestates to the Control Plane every 10 minutes.
    - Send the new livestates to the Control Plane every 5 seconds during the 10 minutes
- I think this is mainly for the speed to show it on the UI. 
- It would be nice to add an option like `livestateFlushInterval` to change them by users.

**Comment by ffjlabo** [comment]
### Platform Provider; Detector

**The purpose of the component**

- Detector check and send application sync state to the Control Plane
- Currently, There are detectors for each platform provider

**Modification idea**
- piped side
  - Every minute, piped does below
    - List apps by plugin name
    - Get the application sync states from the plugin via gRPC call e.g. `Detect(app)`
    - Send them to the Control Plane
- plugin side
  - Get the liveManifests grouped by deployTarget from livestatestore
  - Calculate the diff between live ones and git ones.
  - Make the app sync state referencing the diff and send it to the Control Plane

**Concern**
- There is a different interval to execute the detectors between terraform and other platforms.
  - other platform: 1 minute
  - terraform: 10 minutes
    - This might be to solve performance problem https://github.com/pipe-cd/pipecd/pull/4181
- It would be nice to add an option like `detectInterval` to change them by users.

**Comment by ffjlabo** [comment]
### Platform Provider; Plan Preview

**The purpose of the component**
- Make Plan Preview results ( the diffs, the plan result) for the apps which have some changes

**Modification idea**
- piped side
  - List commands for Plan Preview from the Control Plane by `commandCheckInterval`
  - Every command,  list apps managed by the piped using repoID related to the commands
  - Every app, extract the apps that have some diffs
  - Get the PlanPreviewResult from the plugin via gRPC call e.g. `BuildPlanPreviewResult(app)`
  - Send all results of the apps to the ControlPlane with a command
- plugin side
  - Get the manifests on the latest commit
  - Get the manifests on the commit used by the latest Deployment
  - Make a PlanPreviewResult using the diffs between the above

**Concern**
- It might be good to extract the apps that have some diffs on the plugin side

**Comment by ffjlabo** [comment]
### Platform Provider; Metrics

The metrics `deployment_status` is affected because it has the label `platform_provider`.
https://github.com/pipe-cd/pipecd/blob/548a804d344d7c5faa70bcb6bd0889aa2a9269af/pkg/app/piped/controller/controllermetrics/metrics.go#L23C1-L40C2

```js
deployment_status{application_id="xxxxxxxx",application_kind="KUBERNETES",application_name="simple",deployment="xxxxxxx",instance="127.0.0.1:9085",job="pipecd-ops",launcher_version="v0.49.2",pipecd_component="piped",piped="xxxxx",piped_version="v0.49.2",platform_provider="kubernetes",project="test",status="DEPLOYMENT_CANCELLED"}
```

**Modification idea**
There are two ideas.

1. Create new metrics like `plugin_deployment_status` 
2. Add a new label `deployTarget` to the `deployment_status`

**Concern**

For idea 1.
- pros
  - Completely create it based on the new data model
- cons
  - We should observe both of the metrics (e.g. create two views for both on grafana)
  - It is hard to check the deployment status from piped < v1 to pipedv1 because they are different metrics.

For idea 2.
- pros
  - We can continue to use it.
- cons
  - We have to modify the query to check both of them like `deployment_status{platform_provider="kubernetes"} or deployment_status{deployTarget="kubernetes"}`

**Comment by ffjlabo** [comment]
### Platform Provider; appconfigreporter

**The purpose of the component**
- Detect and update the app config when it has some changes.
- Send the changed app info to the Control Plane, then the Control Plane updates the app with it.

**Modification idea**
It has logic for checking the difference between the app info in the current app config and the already registered app.
  - https://github.com/pipe-cd/pipecd/blob/4a2334a3c96d31c9bcfcbfae02eb6873dd923cdc/pkg/app/piped/appconfigreporter/appconfigreporter.go#L279-L285
- Before checking, we should decide on the actual `kind` using the steps below.
  - if the kind is `Application`, use the labels["kind"]
  - if the kind is old ones such as `KubernetesApp`, use it

**Concern**
- Nothing for now

**Comment by ffjlabo** [comment]
# Kind; event watcher

**The purpose of the component**
- Event Watcher handles events according to the rule set in the app config.
- When it loads the app config, it calls `config.LoadAplication(repoPath, configRelPath string, appKind model.ApplicationKind)` https://github.com/pipe-cd/pipecd/blob/4a2334a3c96d31c9bcfcbfae02eb6873dd923cdc/pkg/app/piped/planpreview/builder.go#L307

**Modification Idea**
- Modify `config.LoadApplication(repoPath, configRelPath string, appKind model.ApplicationKind)` to `LoadApplication(repoPath, configRelPath string, appKind string)`
- Before loading the app config, we should decide on the actual `kind` as a **string** value because these types are enums.
  - Convert it to the **string**
  - if the kind is `Application`, use the labels["kind"]
  - if the kind is old ones such as `KubernetesApp`, use it

**Comment by ffjlabo** [comment]
# Kind; livestatereporter

**The purpose of the component**
- Livestatereporter sends the application state to the Control Plane
  - https://github.com/pipe-cd/pipecd/blob/4a2334a3c96d31c9bcfcbfae02eb6873dd923cdc/pkg/app/piped/livestatereporter/cloudrun/report.go#L112
- Currently, There are livestatereporters for each platform provider

**Modification Idea**
- Add `plugin_kind` to the `model.ApplicationLiveStateSnapshot`
- Determine it by the application with the rule below
  - if the kind is Application, use the labels["kind"]
  - if the kind is old ones such as KubernetesApp, convert them to type string

```
message ApplicationLiveStateSnapshot {
    reserved 2;
    ...
    ApplicationKind kind = 5 [(validate.rules).enum.defined_only = true];
    ...
    string plugin_kind = 16;
```

**Comment by ffjlabo** [comment]
## Kind; detector

**The purpose of the component**
- Detector check and send application sync state to the Control Plane
- Currently, There are detectors for each platform provider
- It has logic for checking
  - whether the kind in the config is supported
  - whether the kind in the config and the one in the actual app are equal
  - https://github.com/pipe-cd/pipecd/blob/44cfc932b2061f4cfbdbd134c912543919b1119d/pkg/app/piped/driftdetector/cloudrun/detector.go#L300

**Modification idea**
- Before checking it,  we should decide on the actual `kind` as a **string** value because these types are enums.
  - if the kind is `Application`, use the labels["kind"]
  - if the kind is old ones such as `KubernetesApp`, convert them to type string

**Comment by ffjlabo** [comment]
## Kind; notifier

**The purpose of the component**
- Notifier notifies the deployment progress to the slack
- piped uses `kind` to show deployment info on the slack message https://github.com/pipe-cd/pipecd/blob/44cfc932b2061f4cfbdbd134c912543919b1119d/pkg/app/piped/notifier/slack.go#L235

**Modification idea**
-  We should decide on the actual kind as a string value because these types are enums.
  - if the kind is Application, use the labels["kind"]
  - if the kind is old ones such as KubernetesApp, convert them to type string

**Comment by ffjlabo** [comment]
## Kind; Plan preview

**The purpose of the component**
- Make Plan Preview results ( the diffs, the plan result) for the apps which have some changes
- There are three purpose to use `kind`
  1. Determine the logic to make diff result. This is a platform-specific thing. https://github.com/pipe-cd/pipecd/blob/44cfc932b2061f4cfbdbd134c912543919b1119d/pkg/app/piped/planpreview/builder.go#L245
  2. Determine the planner to make plan result. This is also a platform-specific thing. https://github.com/pipe-cd/pipecd/blob/44cfc932b2061f4cfbdbd134c912543919b1119d/pkg/app/piped/planpreview/builder.go#L332

**Comment by ffjlabo** [comment]
## Kind; trigger

- Trigger mainly creates a deployment based on commands.
- There are two purposes for using kind
  1. Add Kind to Deployment https://github.com/pipe-cd/pipecd/blob/44cfc932b2061f4cfbdbd134c912543919b1119d/pkg/app/piped/trigger/deployment.go#L80
  2. Load Application config with Kind https://github.com/pipe-cd/pipecd/blob/44cfc932b2061f4cfbdbd134c912543919b1119d/pkg/app/piped/trigger/trigger.go#L225

**Modification idea**
- For 1, No modification for now. But we need to use labels["kind"] in other logic using `Deployment.Kind`.
- For 2, the same modification as event watcher. ref: https://github.com/pipe-cd/pipecd/issues/5252#issuecomment-2449393524

**Comment by ffjlabo** [comment]
### Platform Provider; Metrics

**The purpose of the component**
The metrics `deployment_status` is affected because it has the label `application_kind`.
https://github.com/pipe-cd/pipecd/blob/548a804d344d7c5faa70bcb6bd0889aa2a9269af/pkg/app/piped/controller/controllermetrics/metrics.go#L23C1-L40C2

```js
deployment_status{application_id="xxxxxxxx",application_kind="KUBERNETES",application_name="simple",deployment="xxxxxxx",instance="127.0.0.1:9085",job="pipecd-ops",launcher_version="v0.49.2",pipecd_component="piped",piped="xxxxx",piped_version="v0.49.2",platform_provider="kubernetes",project="test",status="DEPLOYMENT_CANCELLED"}
```

**Modification idea**
- We should decide on the actual kind as a string value because these types are enums.
  - if the kind is Application, use the labels["kind"]
  - if the kind is old ones such as KubernetesApp, convert them to type string

**Comment by ffjlabo** [comment]
### Kind; scheduler

**The purpose of the component**
- Scheduler controls the plan and deployment flow.
- Currently piped use the kind in it for four purpose
  1. Show log with it https://github.com/pipe-cd/pipecd/blob/44cfc932b2061f4cfbdbd134c912543919b1119d/pkg/app/piped/controller/scheduler.go#L106
  2. Add it to the trace data https://github.com/pipe-cd/pipecd/blob/44cfc932b2061f4cfbdbd134c912543919b1119d/pkg/app/piped/controller/scheduler.go#L278
  3. Decide which rollback executer should be used https://github.com/pipe-cd/pipecd/blob/44cfc932b2061f4cfbdbd134c912543919b1119d/pkg/app/piped/controller/scheduler.go#L439
**Modification idea**
  4. Find platform provider using `HasPlatformProvider` https://github.com/pipe-cd/pipecd/blob/44cfc932b2061f4cfbdbd134c912543919b1119d/pkg/app/piped/controller/scheduler.go#L524

**Modification idea**
- For 1. Add method to get kind with the rule below
  - if the kind is Application, use the labels["kind"]
  - if the kind is old ones such as KubernetesApp, convert them to type string
- For 2. same as for 1
- For 3. We don't need to determine by Kind in plugin architecture. Instead of this, we should determine which plugin is used.
- For 4. create `HasDeployTargets` instead of the `HasPlatformprvoders`

**Comment by khanhtc1202** [comment]
### PlatformProvider; UI

**Mostly related to 2 model applications and deployment**

**Comment by khanhtc1202** [comment]
@hongky-1994 will keep working on the UI update for piped v1

> ### PlatformProvider; UI
> #### Purpose
> **Mostly related to 2 model applications and deployment**
> 
> Used in
> 
> * API call AddApplication, UpdateApplication
> * For showing in Application, Deployment detail page
> 
> The source fetched to used in the drawer is come from the piped configuration
> 
> https://github.com/pipe-cd/pipecd/blob/master/web/src/components/application-form/index.tsx#L58-L80
> 
> #### Modification idea
> * Keep it as in the application and deployment model
> * Also, keep the configuration and parser for pipedv0 config the same as it is (no UI affected around that fetch)
> * Make a new Add Application form/suggestion form that doesn't show any platform provider-related information

**Comment by khanhtc1202** [comment]
Bug reported
https://github.com/pipe-cd/pipecd/issues/5536

**Comment by t-kikuc** [comment]
We should delete this hard coding.
https://github.com/pipe-cd/pipecd/blob/c0e822f54322dc6e73c92925066f49755f3600b1/web/src/components/deployments-detail-page/pipeline/index.tsx#L80

**Comment by khanhtc1202** [comment]
Note: For pipelineStage UI, the current pipelineStage model contains fields that being used by UI to perform logic. Those fields should be deprecated, andthe  controlplane UI needs to support both
- [x] `pipelineStage.visible` for enable/disable rollback stage in pipeline -> change to use `pipelineStage.rollback` and `pipelineStage.status` field for pipedv1 created pipeline <- addressed by https://github.com/pipe-cd/pipecd/pull/5950
- [ ] `pipelineStage.requires` contains pipelineStage ID of stages that need to be finished before other stage -> being used to draw the pipeline graph (ref: https://github.com/pipe-cd/pipecd/blob/master/web/src/components/deployments-detail-page/pipeline/index.tsx#L77-L87 ) cc @hongky-1994 
https://github.com/pipe-cd/pipecd/pull/6054

-> Latest update, we will keep `pipelineStage.requires` as pipedv1 implementation temporary, and we will find a better way later.

**Comment by UtkarshAnandd** [comment]
@khanhtc1202 Hey I am new to this org and this issue excites me to work on this if it is possible can u help me assgi this issue to me ?    
Before working on this i will provide a clear report and  description of the things which i am gonna work on
